_episodes/l1-05-abstractions.md

```diff
@@ -66,130 +169,212 @@ def analyse(input_path):
+with it or has not work on it for a while. This audience is normally future-**you**
+reading past-**you**'s code.
+
+For example, given a point with coordinates `point[0]`, `point[1]`, `point[2]`, what do
+each of them mean? Does `0` corresponds to `x`, `y`, or `z`? Or is it `r`, `theta` or
+`z`? Or something else altogether?
+
+Intelligible code aims to:
+
+- Use descriptive names for variables and functions. The times where names should fit
+within 8 characters, and whole lines be less than 72 are long past.
+- Avoid sources of confusion (What does index `0` reference?)
+- Lessen the cognitive load (Before using `c` don't forget to call the function
+`nothing_to_do_with_c` because of historical implementation detail).
 
-- reading input files is separate from creating complex objects
-- saving output files is separate from running computations
-- objects are created in one go
+Some aspects of code legibility will be sorted - or at least flagged out - by the
+formatters and linters already discussed, but others do require a concious effort by the
+developer.
 
-## A few examples of what **not** to do.
+## Use of global variables
 
-### Mixing reading files and creating objects
+While global variables were common in the past, they have become less and less popular
+due to the problems they might bring to the code. Avoid the use global variables, when
+possible. They make the dataflow complex by essence.
 
-Objects that do need files to be created are hard to create and re-create,
-especially during testing.
+Let's consider the following code:
 
 ```python
-class SectionI
-    def __init__(self, some_array):
-        self.some_array = array
-        # BAD!! Now you need to carry this file around every time you want to
-        # instantiate SectionI
-        self.aaa = read_aaa("somefile.aaa")
+SOME_GLOBAL_VARIABLE = 42
+
+def compute_a(measurements):
+    ...
+    result *= SOME_GLOBAL_VARIABLE
+    return result
 ```
 
-### Mixing computing stuff and IO
+Now the result of `compute_a` has a hidden dependency. It's never clear whether
+calling it twice with the same input `measurements` will yield the same result, as
+`SOME_GLOBAL_VARIABLE` might have changed.
+
+In general make sure that all variables have the most limited scope possible. If
+they're only needed within a single function define them there. Wherever
+possible treat variables with wide scope as **constants** (or make them actual
+constants if your language supports it) so you know they're not being modified
+anywhere.
+
+Having said that, there are situations where global variables are needed, or are just
+unavoidable. A typical example would be accessing a database. That database is just one,
+accessible anywhere in the program, and might change in different places. There is no
+way around that, but just make sure that measures are in place to ensure that the state
+of the database is known at any given time.
+
+
+## Mixing IO and creating objects or making calculations
```

Comment:
Not sure this belongs on same level as previous headings as isn't this an example of separation of concerns?